File: graphedu-ui/src/types/api/system/role.ts
/**
 * 角色管理相关类型定义
 * 对应后端：graphedu/common/models/dto/role.py 和 vo/role.py
 */

// ============================================================================
// 请求 DTO 类型
// ============================================================================

/**
 * 角色查询请求
 */
export interface RoleQueryDTO {
  /** 角色名称（模糊查询） */
  roleName?: string
  /** 角色标识（模糊查询） */
  roleKey?: string
  /** 对照sys_data_status（0正常 1停用） */
  status?: '0' | '1'
  /** 页码 */
  page?: number
  /** 每页数量 */
  size?: number
}

/**
 * 创建角色请求
 */
export interface RoleCreateDTO {
  /** 角色名称 */
  roleName: string
  /** 角色标识（student/teacher/admin） */
  roleKey: string
  /** 显示顺序 */
  roleSort?: number
  /** 数据范围（1：全部 2：自定 3：本部门 4：本部门及以下 5：仅本人） */
  dataScope?: '1' | '2' | '3' | '4' | '5'
  /** 对照sys_data_status（0正常 1停用） */
  status?: '0' | '1'
  /** 备注 */
  remark?: string
  /** 功能权限ID列表 */
  functionIds?: number[]
}

/**
 * 更新角色请求
 */
export interface RoleUpdateDTO {
  /** 角色ID */
  roleId: number
  /** 角色名称 */
  roleName?: string
  /** 角色标识 */
  roleKey?: string
  /** 显示顺序 */
  roleSort?: number
  /** 数据范围 */
  dataScope?: '1' | '2' | '3' | '4' | '5'
  /** 对照sys_data_status（0正常 1停用） */
  status?: '0' | '1'
  /** 备注 */
  remark?: string
  /** 功能权限ID列表 */
  functionIds?: number[]
}

/**
 * 修改角色状态请求
 */
export interface RoleStatusChangeDTO {
  /** 角色ID */
  roleId: number
  /** 对照sys_data_status（0正常 1停用） */
  status: '0' | '1'
}

/**
 * 修改角色数据权限范围请求
 */
export interface RoleDatascopeChangeDTO {
  /** 角色ID */
  roleId: number
  /** 部门ID列表 */
  deptIds: number[]
  /** 数据范围（1：全部 2：自定 3：本部门 4：本部门及以下 5：仅本人） */
  dataScope: '1' | '2' | '3' | '4' | '5'
}

/**
 * 角色关联的用户查询请求
 */
export interface RoleUserQueryDTO {
  /** 角色ID（精确查询） */
  roleId?: number
  /** 用户账号 */
  userName?: string
  /** 用户昵称 */
  nickName?: string
  /** 用户邮箱 */
  email?: string
  /** 手机号码 */
  phonenumber?: string
  /** 页码 */
  page?: number
  /** 每页数量 */
  size?: number
}

// ============================================================================
// 响应 VO 类型
// ============================================================================

/**
 * 角色列表项
 */
export interface RoleListVO {
  /** 角色ID */
  roleId: number
  /** 角色名称 */
  roleName: string
  /** 角色标识 */
  roleKey: string
  /** 显示顺序 */
  roleSort: number
  /** 数据范围 */
  dataScope: string
  /** 对照sys_data_status（0正常 1停用 2已删除） */
  status: string
  /** 创建者 */
  createBy?: number
  /** 创建时间 */
  createTime?: string
  /** 更新者 */
  updateBy?: number
  /** 更新时间 */
  updateTime?: string
  /** 备注 */
  remark?: string
}

/**
 * 角色详细信息
 */
export interface RoleDetailVO {
  /** 角色ID */
  roleId: number
  /** 角色名称 */
  roleName: string
  /** 角色标识 */
  roleKey: string
  /** 显示顺序 */
  roleSort: number
  /** 数据范围（1：全部 2：自定 3：本部门 4：本部门及以下 5：仅本人） */
  dataScope: string
  /** 对照sys_data_status（0正常 1停用 2已删除） */
  status: string
  /** 创建者 */
  createBy?: number
  /** 创建时间 */
  createTime?: string
  /** 更新者 */
  updateBy?: number
  /** 更新时间 */
  updateTime?: string
  /** 备注 */
  remark?: string
  /** 已分配的功能权限ID列表 */
  functionIds?: number[]
}

/**
 * 角色简要信息
 */
export interface RoleSimpleVO {
  /** 角色ID */
  roleId: number
  /** 角色名称 */
  roleName: string
  /** 角色标识 */
  roleKey: string
  /** 对照sys_data_status（0正常 1停用 2已删除） */
  status: string
}

/**
 * 角色关联部门响应
 */
export interface RoleDeptVO {
  /** 已选中的部门ID列表 */
  checkedIds: number[]
}

// ============================================================================
// 枚举类型
// ============================================================================

/**
 * 数据范围类型：1：全部 2：自定 3：本部门 4：本部门及以下 5：仅本人
 */
export type DataScopeType = '1' | '2' | '3' | '4' | '5'

File: graphedu/common/models/dto/systemv2/role.py
"""角色管理相关 DTO 模型

职责：
1. 定义 API 请求验证的数据结构
2. 为 CRUD 操作设置不同的字段约束
"""

from typing import Literal

from pydantic import Field

from graphedu.common.models.dto.base import DTO, PageQuery

from .user import UserQueryDTO


class RoleQueryDTO(PageQuery):
    """角色查询 DTO

    用于查询角色列表，支持分页和多条件查询

    Attributes:
        role_name: 角色名称（模糊查询）
        role_key: 角色标识（模糊查询）
        status: 角色状态（0正常 1停用）
    """

    role_name: str | None = Field(default=None, description="角色名称（模糊查询）")
    role_key: str | None = Field(default=None, description="角色标识（模糊查询）")
    status: Literal["0", "1"] | None = Field(default=None, description="对照sys_data_status（0正常 1停用）")


class RoleCreateDTO(DTO):
    """创建角色 DTO - 所有必填字段必须有值

    用于创建新角色

    Attributes:
        role_name: 角色名称
        role_key: 角色标识（student/teacher/admin）
        role_sort: 显示顺序
        data_scope: 数据范围（1：全部 2：自定 3：本部门 4：本部门及以下 5：仅本人）
        status: 角色状态（0正常 1停用）
        remark: 备注（可选）
        function_ids: 功能权限 ID 列表（可选）
    """

    role_name: str = Field(description="角色名称", min_length=1, max_length=30)
    role_key: str = Field(description="角色标识（student/teacher/admin）", min_length=1, max_length=100)
    role_sort: int = Field(default=0, description="显示顺序", ge=0)
    data_scope: Literal["1", "2", "3", "4", "5"] = Field(
        default="1", description="数据范围（1：全部 2：自定 3：本部门 4：本部门及以下 5：仅本人）"
    )
    status: Literal["0", "1"] = Field(default="0", description="对照sys_data_status（0正常 1停用）")
    remark: str | None = Field(default=None, description="备注", max_length=500)
    function_ids: list[int] | None = Field(default=None, description="功能权限ID列表")


class RoleUpdateDTO(DTO):
    """更新角色 DTO - 所有字段可选，仅更新提供的字段

    用于更新角色信息

    Attributes:
        role_id: 角色 ID
        role_name: 角色名称（可选）
        role_key: 角色标识（可选）
        role_sort: 显示顺序（可选）
        data_scope: 数据范围（可选）
        status: 角色状态（可选）
        remark: 备注（可选）
        function_ids: 功能权限 ID 列表（可选）
    """

    role_id: int = Field(description="角色ID")
    role_name: str | None = Field(default=None, description="角色名称", max_length=30)
    role_key: str | None = Field(default=None, description="角色标识", max_length=100)
    role_sort: int | None = Field(default=None, description="显示顺序", ge=0)
    data_scope: Literal["1", "2", "3", "4", "5"] | None = Field(
        default=None, description="数据范围（1：全部 2：自定 3：本部门 4：本部门及以下 5：仅本人）"
    )
    status: Literal["0", "1"] | None = Field(default=None, description="对照sys_data_status（0正常 1停用）")
    remark: str | None = Field(default=None, description="备注", max_length=500)
    function_ids: list[int] | None = Field(default=None, description="功能权限ID列表")


class RoleStatusChangeDTO(DTO):
    """修改角色状态 DTO - 仅包含状态字段

    用于启用或停用角色

    Attributes:
        role_id: 角色 ID
        status: 角色状态（0正常 1停用）
    """

    role_id: int = Field(description="角色ID")
    status: Literal["0", "1"] = Field(description="对照sys_data_status（0正常 1停用）")


class RoleDatascopeChangeDTO(DTO):
    """修改角色数据权限范围 DTO

    用于更新角色的数据权限范围

    Attributes:
        role_id: 角色 ID
        dept_ids: 部门 ID 列表
        data_scope: 数据范围（1：全部 2：自定 3：本部门 4：本部门及以下）
    """

    role_id: int = Field(description="角色ID")
    dept_ids: list[int] = Field(description="部门ID列表")
    data_scope: Literal["1", "2", "3", "4"] = Field(
        default="1", description="数据范围（1：全部 2：自定 3：本部门 4：本部门及以下）"
    )


class RoleUserQueryDTO(UserQueryDTO):
    """角色关联的用户查询 DTO

    用于查询某个角色下的所有用户

    Attributes:
        role_id: 角色 ID（精确查询）
    """

    role_id: int | None = Field(default=None, description="角色ID（精确查询）")
